File: CLI/AnalyzeCommand.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Core.Models;
using Core.Models.Source;
using Services.Concrete;
using Services.Concrete.Analyzer;

namespace CLI
{
    internal class AnalyzeCommand : BaseCommand
    {
        public const string COMMAND_NAME = "analyze";
        protected Source _source;
        protected readonly List<string> _demoPaths = new List<string>();
        protected List<string> _availableSources = new List<string>();
        protected bool _forceAnalyze = false;

        public AnalyzeCommand() : base(COMMAND_NAME, @"Analyze demos, data will be available from the GUI.")
        {
            foreach (Source source in Source.Sources)
            {
                if (source.Name != Pov.NAME)
                {
                    _availableSources.Add(source.Name);
                }
            }
        }

        public override async Task Run(string[] args)
        {
            ParseArgs(args);

            if (_demoPaths.Count == 0)
            {
                Console.WriteLine(@"No demo paths provided");
                return;
            }

            CacheService cacheService = new CacheService();

            int currentDemoNumber = 0;
            foreach (string demoPath in _demoPaths)
            {
                Console.WriteLine($@"Retrieving demo {++currentDemoNumber}/{_demoPaths.Count} {demoPath}");
                Demo demo = DemoAnalyzer.ParseDemoHeader(demoPath);
                if (demo == null)
                {
                    Console.WriteLine($@"Invalid demo {demoPath}");
                    continue;
                }

                if (!_forceAnalyze && cacheService.HasDemoInCache(demo.Id))
                {
                    Console.WriteLine($@"Demo {demoPath} already analyzed.");
                    continue;
                }

                try
                {
                    if (_source != null)
                    {
                        demo.Source = _source;
                    }

                    Console.WriteLine($@"Analyzing demo {demoPath}");
                    DemoAnalyzer analyzer = DemoAnalyzer.Factory(demo);
                    demo = await analyzer.AnalyzeDemoAsync(new CancellationTokenSource().Token);
                    await cacheService.WriteDemoDataCache(demo);
                }
                catch (Exception)
                {
                    Console.WriteLine($@"Error while analyzing demo {demoPath}");
                }
            }
        }

        private new void ParseArgs(string[] args)
        {
            base.ParseArgs(args);

            for (int index = 0; index < args.Length; index++)
            {
                string arg = args[index];
                bool isOption = arg.StartsWith("--");
                if (isOption)
                {
                    switch (arg)
                    {
                        case "--source":
                            if (args.Length > index + 1)
                            {
                                index += 1;
                                string sourceName = args[index];
                                _source = Source.Factory(sourceName);
                                if (_source == null)
                                {
                                    Console.WriteLine($@"Invalid source: {sourceName}");
                                    Environment.Exit(1);
                                }
                            }
                            else
                            {
                                Console.WriteLine(@"Missing --source argument value");
                                Environment.Exit(1);
                            }

                            break;
                        case "--force":
                            _forceAnalyze = true;
                            break;
                        default:
                            Console.WriteLine($@"Unknown option {arg}");
                            Environment.Exit(1);
                            break;
                    }
                }
                else
                {
                    bool isDemoFile = arg.EndsWith(".dem");
                    if (isDemoFile)
                    {
                        bool fileExists = File.Exists(arg);
                        if (!fileExists)
                        {
                            Console.WriteLine($@"The file doesn't exists: {arg}");
                            Environment.Exit(1);
                        }

                        if (_demoPaths.Contains(arg))
                        {
                            continue;
                        }

                        _demoPaths.Add(arg);
                    }
                    else
                    {
                        try
                        {
                            if (arg.EndsWith("\""))
                            {
                                arg = arg.Substring(0, arg.Length - 1) + "\\";
                            }

                            string directoryPath = Path.GetFullPath(arg);
                            bool directoryExists = Directory.Exists(directoryPath);
                            if (directoryExists)
                            {
                                string[] files = Directory.GetFiles(directoryPath, "*.dem");
                                foreach (string file in files)
                                {
                                    if (_demoPaths.Contains(file))
                                    {
                                        continue;
                                    }

                                    _demoPaths.Add(file);
                                }
                            }
                            else
                            {
                                Console.WriteLine($@"The directory doesn't exists: {arg}");
                                Environment.Exit(1);
                            }
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine($@"Invalid directory: {ex.Message}");
                            Environment.Exit(1);
                        }
                    }
                }
            }
        }

        public override void PrintHelp()
        {
            Console.WriteLine(GetDescription());
            Console.WriteLine(@"");
            Console.WriteLine($@"Usage: {Program.ExeName} {COMMAND_NAME} demoPaths... [--source] [--force]");
            Console.WriteLine(@"");
            Console.WriteLine(@"Demos path can be either .dem files location or a directory. It can be relative or absolute.");
            Console.WriteLine($@"The --source argument force the analysis logic of the demo analyzer. Available values: [{string.Join(",", _availableSources)}]");
            Console.WriteLine(@"The --force argument force demos analyzes (ignore cached data).");
            Console.WriteLine(@"");
            Console.WriteLine(@"Examples:");
            Console.WriteLine(@"");
            Console.WriteLine(@"Analyze 1 demo:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\demo.dem""");
            Console.WriteLine(@"");
            Console.WriteLine(@"Analyze multiple demos:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\demo.dem"" ""C:\Users\username\Desktop\demo2.dem""");
            Console.WriteLine(@"");
            Console.WriteLine(@"Analyze all demos in a directory using the ESL analyzer and re-analyze demos that have already been analyzed:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\MyFolder"" --source esl --force");
        }
    }
}

File: src/cli/commands/analyze-command.ts
import fs from 'fs-extra';
import { glob } from 'csdm/node/filesystem/glob';
import { Command } from './command';
import { type DemoSource, SupportedDemoSources } from 'csdm/common/types/counter-strike';
import { migrateSettings } from 'csdm/node/settings/migrate-settings';
import { CliClientMessageName } from 'csdm/server/messages/cli-client-message-name';
import { ServerPushMessageName } from 'csdm/server/messages/server-push-message-name';
import { AnalysisStatus } from 'csdm/common/types/analysis-status';
import type { Analysis } from 'csdm/common/types/analysis';
import { getErrorCodeMessage } from 'csdm/cli/get-error-code-message';
import { isErrorCode } from 'csdm/common/is-error-code';

export class AnalyzeCommand extends Command {
  public static Name = 'analyze';
  private readonly demoPaths: string[] = [];
  private analyzePositions = false;
  private forceAnalyze = false;
  private source: DemoSource | undefined = undefined;
  private sourceFlag = '--source';
  private forceFlag = '--force';
  private analyzePositionsFlag = '--analyze-positions';

  public getDescription() {
    return 'Analyze and persist demos into the database.';
  }

  public printHelp() {
    const sourceValues = SupportedDemoSources.join(',');
    console.log(this.getDescription());
    console.log('');
    console.log(
      `Usage: csdm ${AnalyzeCommand.Name} demoPaths... ${this.formatFlagsForHelp([this.sourceFlag, this.forceFlag, this.analyzePositionsFlag])}`,
    );
    console.log('');
    console.log('Demos path can be either a .dem files path or a directory. It can be relative or absolute.');
    console.log('');
    console.log(
      `The ${this.sourceFlag} flag forces the analysis logic of the demo analyzer. Available values: [${sourceValues}]`,
    );
    console.log(`The ${this.forceFlag} flag forces demos analyzes even if they are already in the database.`);
    console.log(
      `The ${this.analyzePositionsFlag} flag indicates to includes players,projectiles... positions in the analysis.`,
    );
    console.log('');
    console.log('Examples:');
    console.log('');
    console.log('Analyze 1 demo:');
    console.log(`    csdm ${AnalyzeCommand.Name} "C:\\Users\\username\\Desktop\\demo.dem"`);
    console.log('');
    console.log('Analyze multiple demos:');
    console.log(
      `    csdm ${AnalyzeCommand.Name} "C:\\Users\\username\\Desktop\\demo.dem" "C:\\Users\\username\\Desktop\\demo2.dem"`,
    );
    console.log('');
    console.log(
      'Analyze all demos in a directory using the ESL analyzer and re-analyze demos that have already been analyzed:',
    );
    console.log(
      `    csdm ${AnalyzeCommand.Name} "C:\\Users\\username\\Desktop\\MyFolder" ${this.sourceFlag} esl ${this.forceFlag}`,
    );
  }

  public async run() {
    await this.parseArgs();

    if (this.demoPaths.length === 0) {
      console.log('No demos found');
      this.exitWithFailure();
    }

    await migrateSettings();
    const client = await this.connectToDaemon();

    console.log(`${this.demoPaths.length} demos to process`);

    const pendingChecksums = new Set<string>();
    const lastStatusPerChecksum = new Map<string, AnalysisStatus>();
    let hasError = false;
    let resolveCompletion: () => void;
    const completion = new Promise<void>((resolve) => {
      resolveCompletion = resolve;
    });

    const markAnalysisAsDone = (checksum: string) => {
      pendingChecksums.delete(checksum);
      if (pendingChecksums.size === 0) {
        resolveCompletion();
      }
    };

    const onAnalysisUpdated = (analysis: Analysis) => {
      const { demoChecksum: checksum, demoPath, status } = analysis;
      if (!pendingChecksums.has(checksum) || lastStatusPerChecksum.get(checksum) === status) {
        return;
      }
      lastStatusPerChecksum.set(checksum, status);

      switch (status) {
        case AnalysisStatus.Analyzing:
          console.log(`Analyzing demo ${demoPath}...`);
          break;
        case AnalysisStatus.Inserting:
          console.log(`Inserting match into database ${demoPath}...`);
          break;
        case AnalysisStatus.InsertSuccess:
          console.log(`Demo ${demoPath} inserted into the database`);
          markAnalysisAsDone(checksum);
          break;
        case AnalysisStatus.AnalyzeError:
        case AnalysisStatus.InsertError:
          hasError = true;
          console.error(
            status === AnalysisStatus.AnalyzeError
              ? `Error analyzing demo ${demoPath}`
              : `Error inserting match into database ${demoPath}`,
          );
          if (analysis.output !== '') {
            console.error(analysis.output);
          }
          markAnalysisAsDone(checksum);
          break;
      }
    };

    client.on(ServerPushMessageName.AnalysisUpdated, onAnalysisUpdated);

    const { addedDemos, skippedDemoPaths } = await client.send(
      {
        name: CliClientMessageName.AddDemoPathsToAnalyses,
        payload: {
          demoPaths: this.demoPaths,
          force: this.forceAnalyze,
          analyzePositions: this.analyzePositions,
          source: this.source,
        },
      },
      { timeoutMs: 20_000 },
    );

    for (const demoPath of skippedDemoPaths) {
      console.log(`Demo ${demoPath} already in database, skipping this demo.`);
    }
    for (const { checksum } of addedDemos) {
      pendingChecksums.add(checksum);
    }

    if (pendingChecksums.size > 0) {
      const daemonStatusPollIntervalMs = 30_000;
      // Safety net in case a terminal push message never arrives (e.g. the analysis has been removed from the queue
      // through the GUI). Push messages and the status reply arrive on the same socket, so a non-busy status with
      // pending analyses means they will never complete.
      const pollIntervalId = setInterval(async () => {
        try {
          const daemon = await client.send({ name: CliClientMessageName.GetDaemonStatus });
          if (!daemon.busy && pendingChecksums.size > 0) {
            hasError = true;
            console.error('Some analyses did not complete, check the demos in the GUI or re-run the command.');
            resolveCompletion();
          }
        } catch (error) {
          hasError = true;
          let errorMessage: string;
          if (isErrorCode(error)) {
            errorMessage = getErrorCodeMessage(error);
          } else {
            errorMessage = error instanceof Error ? error.message : 'The daemon is not responding.';
          }
          console.error(errorMessage);
          resolveCompletion();
        }
      }, daemonStatusPollIntervalMs);

      await completion;
      clearInterval(pollIntervalId);
    }

    client.close();

    if (hasError) {
      this.exitWithFailure();
    }
  }

  protected async parseArgs() {
    super.parseArgs(this.args);

    if (this.args.length === 0) {
      console.log('No demo path provided');
      this.printHelp();
      this.exitWithFailure();
    }

    for (let index = 0; index < this.args.length; index++) {
      const arg = this.args[index];
      if (this.isFlagArgument(arg)) {
        switch (arg) {
          case this.analyzePositionsFlag:
            this.analyzePositions = true;
            break;
          case this.forceFlag:
            this.forceAnalyze = true;
            break;
          case this.sourceFlag:
            if (this.args.length > index + 1) {
              index += 1;
              const source = this.args[index] as DemoSource;
              const isValidSource = SupportedDemoSources.includes(source);
              if (!isValidSource) {
                console.log(`Invalid source ${source}`);
                this.exitWithFailure();
              }
              this.source = source;
            } else {
              console.log(`Missing ${this.sourceFlag} value`);
              this.exitWithFailure();
            }
            break;
          default:
            console.log(`Unknown flag: ${arg}`);
            this.exitWithFailure();
        }
      } else {
        try {
          const stats = await fs.stat(arg);
          if (stats.isDirectory()) {
            const files = await glob('*.dem', {
              cwd: arg,
              absolute: true,
            });
            this.demoPaths.push(...files);
          } else if (stats.isFile() && arg.endsWith('.dem')) {
            this.demoPaths.push(arg);
          } else {
            console.log(`Invalid path: ${arg}`);
            this.exitWithFailure();
          }
        } catch (error) {
          console.log(`Invalid path: ${arg}`);
          this.exitWithFailure();
        }
      }
    }
  }
}
